### PR TITLE
chez-racket: init at 2021-12-11

### PR DESCRIPTION
**Pull request by L-as**
###### Motivation for this change

This is basically Chez Scheme but better, notably it supports aarch64-linux (and likely aarch64-darwin too!).
This can be used to make Idris 2 work on more platforms with its Chez Scheme backend, albeit it is not done in this PR.

###### Things done

I've added `chez-racket` as a separate directory with its own files, since it's a permanent fork.
The only common part is the setup hook.

It would be great if someone on `aarch64-darwin` could test this.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 131833` at 1ad518f5 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>chez-racket</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md">stale-substitute</a>

Stale substituteInPlace detected.
substituteStream(): WARNING: pattern '/bin/ln' doesn't match anything in file './workarea'
Near pkgs/development/compilers/chez-racket/default.nix:52:0:
```
   |
52 |     description  = "Fork of Chez Scheme for Racket";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md">stale-substitute</a>

Stale substituteInPlace detected.
substituteStream(): WARNING: pattern '/bin/cp' doesn't match anything in file './workarea'
Near pkgs/development/compilers/chez-racket/default.nix:52:0:
```
   |
52 |     description  = "Fork of Chez Scheme for Racket";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/patch-phase.md">patch-phase</a>

`patchPhase` should not be overridden, use `postPatch` instead.

Near pkgs/development/compilers/chez-racket/default.nix:25:3:
```
   |
25 |   patchPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/chez-racket/default.nix:43:3:
```
   |
43 |   configurePhase = ''
   |   ^
```
</li>
  </ul>
</details>

File: pkgs/development/compilers/chez-racket/default.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
+{ stdenv, lib, fetchFromGitHub
+, coreutils, cctools
+, ncurses, libiconv, libX11
+}:
+
+stdenv.mkDerivation rec {
+  pname = "chez-scheme-racket";
+  version = "2021-07-09";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/chez-racket/default.nix:8]:
```suggestion
  version = "unstable-2021-07-09";
```

File: pkgs/development/compilers/chez-racket/default.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  src = fetchFromGitHub {
+    owner  = "racket";
+    repo   = "ChezScheme";
+    rev    = "f1f2294be7809bbc15c05af6f7a99e2b47070c18";
+    sha256 = "wXjPyTKqIifRKoeBOn+G+pq9QK/1suTe5VqqZNanSOI=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [ coreutils ] ++ lib.optional stdenv.isDarwin cctools;
+  buildInputs = [ ncurses libiconv libX11 ];
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/chez-racket/default.nix:19]:
```suggestion
  buildInputs = [ ncurses libX11 ]
    ++ lib.optional stdenv.isDarwin libiconv;
```

**Comment by L-as** [comment]
Bump

File: pkgs/development/compilers/chez-racket/default.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+stdenv.mkDerivation rec {
+  pname = "chez-scheme-racket";
+  version = "unstable-2021-07-09";
+
+  src = fetchFromGitHub {
+    owner  = "racket";
+    repo   = "ChezScheme";
+    rev    = "f1f2294be7809bbc15c05af6f7a99e2b47070c18";
+    sha256 = "wXjPyTKqIifRKoeBOn+G+pq9QK/1suTe5VqqZNanSOI=";
+    fetchSubmodules = true;
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/chez-racket/default.nix:15]: We don't want to use the vendored lz4 or zlib and preferable none of the other libs, too.

Comment (L-as) [reply pkgs/development/compilers/chez-racket/default.nix:15]: I took a look at it and it does not seem trivial unfortunately.

Comment (LiberalArtist) [reply pkgs/development/compilers/chez-racket/default.nix:15]: You can pass `configure` your shared zlib and lz4: https://github.com/racket/ChezScheme/blob/91a33d86e8e5f637bbc96afa765aff6b3b66120c/configure#L320-L332

Nanopass and STeX are not really shared libraries and are only used in the build process, but in Guix we share their origin definitions with the upstream/Cisco Chez Scheme package.

Comment (L-as) [reply pkgs/development/compilers/chez-racket/default.nix:15]: Thanks for the help!

File: pkgs/development/compilers/chez-racket/default.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  # Don't use configureFlags, since that just implicitly appends
+  # everything onto a --prefix flag, which ./configure gets very angry
+  # about.
+  #
+  # NB: Cross-compilation won't work if we're using `make auto.bootquick`,
+  # since that targets the build platform.
+  configurePhase = ''
+    runHook preConfigure
+    ./configure --pb
+    make auto.bootquick
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/chez-racket/default.nix:36]:
```suggestion
    make auto.bootquick -J$NIX_BUILD_CORES
```

File: pkgs/development/compilers/chez-racket/default.nix [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  src = fetchFromGitHub {
+    owner  = "racket";
+    repo   = "ChezScheme";
+    rev    = "f1f2294be7809bbc15c05af6f7a99e2b47070c18";
+    sha256 = "wXjPyTKqIifRKoeBOn+G+pq9QK/1suTe5VqqZNanSOI=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [ coreutils ]
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/chez-racket/default.nix:18]: coreutils is already part of the stdenv and does not need to be listed here.

**Comment by L-as** [comment]
I've fixed all the problems and in the process also added support for cross-compilation!

**Comment by Steven0351** [comment]
Result of `nixpkgs-review pr 131833` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>chez-racket</li>
  </ul>
</details>

**Comment by Steven0351** [comment]
Package builds for aarch64-darwin but does not run:
```
[nix-shell:~/.cache/nixpkgs-review/pr-131833]$ ./results/chez-racket/bin/scheme
Killed: 9
```

**Comment by L-as** [comment]
While I'm not able to test this myself, it seems to me that the chance that this is a bug upstream is somewhat high.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 131833` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>chez-racket</li>
  </ul>
</details>

File: pkgs/development/compilers/chez-racket/shared.nix [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  prePatch = ''
+    rm -rf zlib/*.c lz4/lib/*.c
+  '';
+
+  postPatch = ''
+    export ZLIB="$(find ${zlib.out}/lib -type f | sort | head -n1)"
+    export LZ4="$(find ${lz4.out}/lib -type f | sort | head -n1)"
+  '';
+
+  nativeBuildInputs = lib.optional stdenv.isDarwin darwin.cctools;
```
Comment (Steven0351) [inline pkgs/development/compilers/chez-racket/shared.nix:26]: I was able to get this working on aarch64-darwin by changing this to
```nix
nativeBuildInputs = lib.optional stdenv.isDarwin [darwin.cctools darwin.autoSignDarwinBinariesHook];
```

Ran the REPL and voila:
```
[nix-shell:~/.cache/nixpkgs-review/rev-a9f68fe379318d19830db01846db215359bb8598]$ ./results/chez-racket/bin/scheme
Chez Scheme Version 9.5.7.2
Copyright 1984-2021 Cisco Systems, Inc.

> (+ 3 4)
7
> (exit)
```

Comment (Steven0351) [reply pkgs/development/compilers/chez-racket/shared.nix:26]: I got the hint for this fix from https://github.com/NixOS/nixpkgs/issues/149692#issuecomment-989086581

Comment (L-as) [reply pkgs/development/compilers/chez-racket/shared.nix:26]: Thanks! I'll commit this.

**Comment by Steven0351** [comment]
Result of `nixpkgs-review pr 131833` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>chez-racket</li>
  </ul>
</details>

**Comment by Steven0351** [comment]
Result of `nixpkgs-review pr 131833` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>chez-racket</li>
  </ul>
</details>

**Review by Steven0351 — APPROVED** [review]

**Comment by Steven0351** [comment]
Everything builds and runs on x86_64-darwin and aarch64-darwin

**Review by siraben — APPROVED** [review]
LGTM
```ShellSession
[nix-shell:~/.../nixpkgs-review/pr-131833-1]$ scheme
Chez Scheme Version 9.5.7.2
Copyright 1984-2021 Cisco Systems, Inc.

> (define (fact n) (if (= n 0) 1 (* n (fact (- n 1)))))
> (fact 6)
720
```

**Comment by siraben** [comment]
It appears all four platforms work, so I'm going to merge this.

**Pull request merged**